src/Exec.hpp

```diff
@@ -1,30 +1,52 @@
 #ifndef EXEC_HPP
 #define EXEC_HPP
 
+#include "stddef.h"
+#include <thread>
+
 namespace t2
 {
+  struct NodeData;
+  struct MemAllocHeap;
+  struct BuildQueue;
+
   struct EnvVariable
   {
     const char *m_Name;
     const char *m_Value;
   };
 
+  struct OutputBufferData
+  {
+     char* buffer;
+     size_t buffer_size;
+     int cursor;
+     MemAllocHeap* heap;
+  };
+
   struct ExecResult
   {
-    int     m_ReturnCode;
-    bool    m_WasSignalled;
+    int               m_ReturnCode;
+    bool              m_WasSignalled;
+    NodeData*         m_FrozenNodeData;
+    OutputBufferData  m_StdOutBuffer;
+    OutputBufferData  m_StdErrBuffer;
   };
 
-  void ExecInit(void);
+  void InitOutputBuffer(OutputBufferData* data, MemAllocHeap* heap);
+  void ExecResultFreeMemory(ExecResult* result);
+  void ExecInit();
 
   ExecResult ExecuteProcess(
         const char*         cmd_line,
         int                 env_count,
         const EnvVariable*  env_vars,
+        MemAllocHeap*       heap,
         int                 job_id,
-        int                 echo_cmdline,
-        const char*         annotation,
-        bool                force_use_tty = false);
+        bool                stream_output_to_stdout,
+        std::function<int()>* callback_on_slow,
```

Comment:
I'm not super familiar with modern C++, but I've heard things like "std::function allocates memory" or somesuch. Given how careful Tundra is about memory management, would be good to check whether this particular usage does not allocate.

Reply:
`std::function` is virtually guaranteed to allocate heap memory.